compiler/optimizations/forallOptimizations.cpp

```diff
@@ -46,12 +47,39 @@
 //                          applicable last statements within `forall` bodies
 
 static int curLogDepth = 0;
-static void LOG_ALA(int depth, const char *msg, BaseAST *node);
+static bool LOG_ALA(int depth, const char *msg, BaseAST *node);
 static void LOGLN_ALA(BaseAST *node);
 
-static void LOG_AA(int depth, const char *msg, BaseAST *node);
+static bool LOG_AA(int depth, const char *msg, BaseAST *node);
 static void LOGLN_AA(BaseAST *node);
 
+// Support for reporting calls that are not optimized for different reasons
+enum CallRejectReason {
+  CRR_ACCEPT,
+  CRR_NOT_ARRAY_ACCESS_LIKE,
+  CRR_NO_CLEAN_INDEX_MATCH,
+  CRR_ACCESS_BASE_IS_LOOP_INDEX,
+  CRR_ACCESS_BASE_IS_NOT_OUTER_VAR,
+  CRR_ACCESS_BASE_IS_SHADOW_VAR,
+  CRR_TIGHTER_LOCALITY_DOMINATOR,
+  CRR_UNKNOWN,
+};
+
+struct ConstCharCmp {
+  bool operator() (const char *lhs, const char *rhs) const {
+    return strcmp(lhs, rhs) < 0;
+  }
+};
+
+// we store all the locations where we have added these primitives. When we
+// report finalizing an optimization (either positively or negatively) we remove
+// those locations from the set. Towards the end of resolution, if there are
+// still items left in these sets, those indicate that we reverted the
+// optimization by completely removing the block that those primitives are in.
+std::set<const char *, ConstCharCmp> primMaybeLocalThisLocations;
+std::set<const char *, ConstCharCmp> primMaybeAggregateAssignLocations;
```

Comment:
In the last two commits

- I moved the stringLoc logic from BaseAST to astlocT to avoid having to generate a location string here
- And replaced the key type of these sets with astlocT

However, I think there's still some complexity here. I still want to use `ConstCharCmp` or something equivalent to be able to store locations in order, so that when we print them out they are in a deterministic and a reasonable order. I believe that this is still the right direction, though. But as a future step we might want to store a `fileno` in `astlocT` and add comparators to that type to be able to sort AST locations.

Does that make sense to you?

Reply:
If we don't already have `<` etc for `astlocT` then we should definately add them.

Reply:
(The other steps are OK with me - I think you should add the `astlocT` comparison operators in this PR though).

Reply:
Added those operators, and updated the PR message. Running more testing.